feat: Added support for file uploads.

yie1d opened the pull request:
### **User description**
### Description  

Added support for file uploads by setFileInputFiles and listening filechooser events    

### Changes  
- Add function `set_input_files` for `WebElement`, which supports elements like `<input type="file">`.
- Add context manager `expect_file_chooser` for `Page`, which supports other file chooser dialogs.

### Usage  
Elements like `<input type="file">`:  
```python
file_input_field = await page.find_element(By.XPATH, '//input[@type="file"]')

# Single-file upload
# await file_input_field.set_input_files(r'c:\demo\demo1.file')
# Multi-file uploads
await file_input_field.set_input_files([r'c:\demo\demo1.file', r'c:\demo\demo2.file'])
```

Others:
```python
async with page.expect_file_chooser(files=r'c:\demo\demo1.file'):
    file_input_field = await page.find_element(By.XPATH, '//button[@id="upload-button-demo"]')
    await file_input_field.click()
```

### Testing  
All tests passing.


___

### **PR Type**
Enhancement


___

### **Description**
- Added support for file uploads via `set_input_files` and `expect_file_chooser`.

- Introduced `FileUploadCommands` for handling file input operations.

- Added context manager for intercepting file chooser dialogs.

- Enhanced error handling for file input validation.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>page.py</strong><dd><code>Add file chooser dialog interception and handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pydoll/browser/page.py

<li>Added <code>expect_file_chooser</code> context manager for file chooser dialogs.<br> <li> Introduced <code>set_intercept_file_chooser_dialog</code> for enabling/disabling <br>interception.<br> <li> Added property and methods to manage file chooser dialog interception.


</details>


  </td>
  <td><a href="https://github.com/thalissonvs/pydoll/pull/77/files#diff-b2e0fc67c9966d15536885887f6d1e8e5889a640f31431d2ebf421d5625b724d">+72/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>file_upload.py</strong><dd><code>Introduce `FileUploadCommands` for file upload handling</code>&nbsp; &nbsp; </dd></summary>
<hr>

pydoll/commands/file_upload.py

<li>Created <code>FileUploadCommands</code> for file upload operations.<br> <li> Added file existence validation for upload commands.<br> <li> Implemented methods to generate CDP commands for file uploads.


</details>


  </td>
  <td><a href="https://github.com/thalissonvs/pydoll/pull/77/files#diff-0a4d79173a798b79f8f05a5d80c5d0980d8d195984ea27219fec8efd4f5c4697">+62/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>page.py</strong><dd><code>Extend `PageCommands` with file chooser interception</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pydoll/commands/page.py

<li>Added <code>set_intercept_file_chooser_dialog</code> command for file chooser <br>interception.<br> <li> Enhanced <code>PageCommands</code> with new file chooser-related methods.


</details>


  </td>
  <td><a href="https://github.com/thalissonvs/pydoll/pull/77/files#diff-46adff4f3a44b078e4fb75890c7149dc5f6cbe9a138a7f52f26b08d82db909ae">+19/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>element.py</strong><dd><code>Add `set_input_files` for file input elements</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pydoll/element.py

<li>Added <code>set_input_files</code> method for file input elements.<br> <li> Validated element type before setting file inputs.<br> <li> Integrated <code>FileUploadCommands</code> for file input handling.


</details>


  </td>
  <td><a href="https://github.com/thalissonvs/pydoll/pull/77/files#diff-0321a2e5286c551bc1882ec9e165a23ff842589c0ab6a9a4ddd0984d4b75821f">+15/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

thalissonvs commented on pydoll/commands/file_upload.py:
```diff
@@ -0,0 +1,63 @@
+import copy
```
You can put this code in the `dom commands` file, since it belongs to the same domain

thalissonvs commented on pydoll/browser/page.py:
```diff
@@ -72,6 +78,16 @@ def dom_events_enabled(self) -> bool:
         return self._dom_events_enabled
+    
+    @property
+    def intercept_file_chooser_dialog(self) -> bool:
```
```suggestion
    def intercept_file_chooser_dialog_enabled(self) -> bool:
```

just to maintain the same pattern :)

thalissonvs commented on pydoll/commands/file_upload.py:
```diff
@@ -0,0 +1,63 @@
+        Returns:
+            dict: The CDP command to set the file input files.
+        """
+        print(type(backend_node_id))
```
I think you forgot this print

thalissonvs replied:
I think you forgot this print

thalissonvs commented:
Hi @yie1d, thank you so much for your contribution!

I left some comments to maintain the patterns used throughout the code.
Also, you need to run the code linter using poetry run task lint and write tests for the new functionalities you've added.

If you have any questions, feel free to get in touch.

yie1d commented:
> Hi @yie1d, thank you so much for your contribution!
> 
> I left some comments to maintain the patterns used throughout the code. Also, you need to run the code linter using poetry run task lint and write tests for the new functionalities you've added.
> 
> If you have any questions, feel free to get in touch.


### **Updates Implemented**  
  
Hi @thalissonvs ,I've addressed all your feedback in the latest commit.  

### **Testing**    

- Added tests for `upload_files`  
- Added tests for `disable_intercept_file_chooser_dialog`  
- Added tests for `expect_file_chooser` 

*tips:*  
`upload_files` has file existence validation during uploads，implemented `missing_ok=True` in test cleanup logic to suppress missing file errors  


Feel free to review the changes, and if there are any further issues or edge cases I missed, please let me know! Happy to iterate

thalissonvs commented:
Hi @yie1d , how are you?
The code is perfect, thanks.
But the tests and lint CI are still failing. Do you need any help to fix them? 
Also, I've created a CONTRIBUTING.md file, so you can check if you have any doubt

byayex commented:
Any Update?

yie1d commented:
> Any Update?

All feedback addressed! Awaiting final review from reviewers.

yie1d commented:
> Do we have any method to wait for the file/files to be fully uploaded?


We can generally determine whether the file upload progress bar is exists or the element contains success/failure attribute.   
For example:
```python
progress_bar = await page.find_element(By.XPATH, '//*[@id="upload-demo-progress"]', raise_exc=False)
if progress_bar:
    print('progress_bar is exists, file uploading...')
else:
    print('progress_bar is not exists, the file is successfully uploaded!')
```
You can continue retrying as long as the progress bar remains visible, until it disappears.

thalissonvs commented:
Hi @yie1d , thank you so much! I'll merge now
